server/util/flagutil/types/autoflags/tags/tags.go

```diff
@@ -0,0 +1,199 @@
+package tags
+
+import (
+	"flag"
+	"fmt"
+
+	"github.com/buildbuddy-io/buildbuddy/server/util/flagutil/common"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+)
+
+type Tagged interface {
+	Value() flag.Value
+	Zero() string
+
+	// from flag.Value
+	SetFunc(func(value string) error)
+	StringFunc(func() string)
+
+	// from WrappingValue
+	WrappedValueFunc(func() flag.Value)
+
+	// from Expandable
+	ExpandFunc(func(mapping func(string) (string, error)) error)
+
+	// from Secretable
+	IsSecretFunc(func() bool)
+
+	// from SetValueForFlagNameHooked
+	SetValueForFlagNameHookFunc(func())
+
+	// from YAMLSetValueHooked
+	YAMLSetValueHookFunc(func())
+}
+
+type Taggable interface {
+	Tag(flagset *flag.FlagSet, name string, f Tagged)
+}
+
+type TagFlag[T any] struct {
+	value flag.Value
+
+	setFunc                     func(value string) error
+	stringFunc                  func() string
+	wrappedValueFunc            func() flag.Value
+	expandFunc                  func(mapping func(string) (string, error)) error
+	isSecretFunc                func() bool
+	setValueForFlagNameHookFunc func()
+	yamlSetValueHookFunc        func()
+}
+
+func (t *TagFlag[T]) Value() flag.Value {
+	return t.value
+}
+
+func (_ *TagFlag[T]) Zero() string {
+	return fmt.Sprint(common.Zero[T]())
+}
+
+func (t *TagFlag[T]) SetFunc(setFunc func(string) error) {
+	t.setFunc = setFunc
+}
+
+func (t *TagFlag[T]) Set(value string) error {
+	if t.setFunc != nil {
+		return t.setFunc(value)
+	}
+	return t.Value().Set(value)
+}
+
+func (t *TagFlag[T]) StringFunc(stringFunc func() string) {
+	t.stringFunc = stringFunc
+}
```

Comment:
maybe prefix the setters with `Set`? I realize `SetSetFunc` is awkward but I think it's better overall for the setters to be named more intuitively. Another option is to replace `*Func` with `-er`, like `ValueSetter`, `Expander`, etc.